src/librustc_codegen_llvm/mir/constant.rs

```diff
@@ -68,7 +70,7 @@ pub fn primval_to_llvm(cx: &CodegenCx,
                 &C_usize(cx, ptr.offset.bytes()),
                 1,
             ) };
-            if scalar.value != layout::Pointer {
+            if layout.value != layout::Pointer {
                 unsafe { llvm::LLVMConstPtrToInt(llval, llty.to_ref()) }
             } else {
                 consts::bitcast(llval, llty)
```

Comment:
Looking at this, this sort of thing probably needs to always use `bitcast` after `LLVMConstPtrToInt`, in case the latter is needed at all. You should add a test for an `union` used to convert a reference into a `f32` / `f64` (depending on `#[cfg(target_pointer_width)]`), which I highly suspect trips a LLVM assertion in the current state of the code.